src/ess/reduce/ui.py

```diff
@@ -232,3 +234,136 @@ def refresh_workflow_box(change) -> None:
+            node, new_parameter_values.get(parameter_box._input_registry[node].name, {})
+        )
+        # Parse the new values and corresponding fields in the existing widget
+        new_fields: dict[str, Any] = new_internal_values.get('fields', {})
+        new_fields_keys = set(new_fields.keys())
+        widget_fields = getattr(widget, "fields", {})
+        # Extract valid fields
+        valid_fields = new_fields_keys & set(widget_fields.keys())
+        # Warn for invalid fields
+        invalid_fields = new_fields_keys - valid_fields
+        for field_name in invalid_fields:
+            warning_msg = f"Cannot set field '{field_name}' for parameter '{node}'."
+            " The field does not exist in the widget. "
+            "The field value will be ignored."
+            warnings.warn(warning_msg, UserWarning, stacklevel=1)
+        # Set the valid fields
+        for field_name in valid_fields:
+            widget_fields[field_name].value = new_fields[field_name]
+
+        # Set the high-level value if needed and possible.
+        # It is set after the fields to prioritize the high-level value.
+        if (self_value := new_internal_values.get('value')) is not None:
+            if _has_widget_value_setter(widget):
+                widget.value = self_value
+            else:
+                warning_msg = f"Cannot set value for parameter '{node}'. "
+                "The widget does not have a value setter. The value will be ignored."
+                warnings.warn(warning_msg, UserWarning, stacklevel=1)
+
+
+def _is_value_exportable(value: Any) -> bool:
+    return isinstance(value, int | float | str | bool)
+
+
+def get_input_widget_values(
+    widget: WorkflowWidget | ParameterBox, key_as_str: bool = False
+) -> dict[str | type, dict[str, Any]]:
+    """Return the current values of the input widgets in the target widget.
+
+    Empty fields and widgets without value setters will not be included in the result.
+    The result of this function can be used to set the values of the input widgets
+    using the `~set_input_widget_values` function.
+
+    Parameters
+    ----------
+    widget:
+        The widget containing the input widgets.
+    key_as_str:
+        If True, the keys in the returned dictionary will be the parameter names(str).
+        Otherwise, the keys will be type objects(Key in Sciline graphs).
+
+    """
+    parameter_box = _get_parameter_box(widget)
+    parameter_input_widget_values = {}
+    for node, widget in parameter_box._input_widgets.items():
+        input_widget_values = {}
+        fields_values = {
+            field_name: field.value
+            for field_name, field in getattr(widget, 'fields', {}).items()
+        }
+        if fields_values:  # Skip setting empty fields
+            input_widget_values['fields'] = fields_values
+        if _has_widget_value_setter(widget) and _is_value_exportable(widget.value):
+            input_widget_values['value'] = widget.value
```

Comment:
I expected them to be stored in a json file so limited types of values that can be stored here.

Reply:
Have you considered letting the code storing into JSON deal with that? If you "handle" this here it will also "break" all other applications which do not store to JSON.

Reply:
> it will also "break" all other applications which do not store to JSON.

I'm not sure what you mean by this.
In the updated implementation, the type of the value is not handled anymore.